Retry on play fail and no connection

mattcarter11 opened the pull request:
#### What is it?
- [ ] New feature (user facing)
- [x] Update to existing feature (user facing)
- [ ] Bugfix (user facing)
- [ ] Translations
- [ ] Codebase improvements or refactors (dev facing)
- [ ] Other
<!-- If you selected Translations, please this pull request only contains tranlations changes, and NO code changes (unless it is absolutely nessesary, please provide a description of why). Please ensure your are able to build and run the app. You can simply delete the next 4 sections as they do not apply. -->

#### Description of the changes in your PR
<!-- While bullet points are the norm in this section, feel free to write free-form text instead of a list -->
- add logic to handle error when no connection
- expose flag when player is waiting for connection
- add logic to resume playing on connection
- add error message when this happens

Closes #181 

#### Due diligence
- [x] I read the [contribution guidelines](../CONTRIBUTING.md).
- [x] I understand that in the event of merge conflicts, **I may be asked to rebase my branch** on top of the dev branch, instead of resolving them by merging dev into my own branch

mattcarter11 commented:
@mikooomich conflicts fixed

mikooomich commented on app/src/main/java/com/dd3boh/outertune/constants/PreferenceKeys.kt:
```diff
@@ -46,7 +46,6 @@ val AudioOffload = booleanPreferencesKey("enableOffload")
 
 val PersistentQueueKey = booleanPreferencesKey("persistentQueue")
 val SkipSilenceKey = booleanPreferencesKey("skipSilence")
-val SkipOnErrorKey = booleanPreferencesKey("skipOnError")
```
Why remove this key? skipOnError on: auto skip is on. skipOnError off: pause on error

mikooomich commented on app/src/main/java/com/dd3boh/outertune/playback/MusicService.kt:
```diff
@@ -213,6 +216,20 @@ class MusicService : MediaLibraryService(),
             }
         }
 
+        networkConnectivityObserver = NetworkConnectivityObserver(this)
```
Should probably use the same connectivity observer from main activity?

mattcarter11 replied:
how?

mikooomich commented on app/src/main/java/com/dd3boh/outertune/utils/Retry.kt:
```diff
@@ -0,0 +1,55 @@
+            return block()
+        } catch (e: Throwable) {
+            val nextDelay = currentDelay + Random.nextLong(0, jitterRange)
+            Timber.tag("Retry").d("$functionName attempt ${attempt + 1} of $times failed, retrying in $nextDelay ms")
```
I think tagging with "NetworkUtils" or something would be a better tag

mikooomich commented on app/src/main/res/values/strings.xml:
```diff
@@ -207,6 +207,9 @@
     <string name="error_timeout">Timeout</string>
     <string name="error_unknown">Unknown error</string>
+    <string name="on_error">On error</string>
+    <string name="pause">Pause</string>
```
Both the notification stuff in music service and these strings are removed or moved to another file respectively. I'll can fix these 2 and then just squash merge in the end.


As always, thanks for your work!

mattcarter11 replied:
Ok, move them where they should be

mikooomich committed:
[squash rebase] player: when playback error because no connection, stop song and show loading. When connection back, start playing again

onError: finish wait to reconnect

rebase: fix

on error: simplify logic (remove toggle, just wait for connection or skip upt to two songs on error)

refactor: remove duplicate enum

player: Fixup new on-error logic

* Resurrect runaway diesel engine protection for auto skip
* Re-order if else such to show error message on auto skip
* Separate wait to reconnect on it's own

retry: if pause stop wait for reconnect

update to new string changes

mikooomich committed:
thumbnail size
